File: python/pascals-triangle/pascals_triangle.py
def rows(row_count):

    if row_count < 0:
        raise ValueError("number of rows is negative")

    if row_count == 0:
        return []

    if row_count == 1:
        return [[1]]

    all_rows = rows(row_count - 1)
    all_rows.append(generate_row(all_rows[-1]))

    return all_rows


def generate_row(previous_row):

    row0 = list(previous_row)
    row0.append(0)

    row1 = [0]
    row1.extend(previous_row)

    return [a + b for (a, b) in (zip(row0, row1))]

File: julia/pascals-triangle/pascals-triangle.jl
function triangle(n)

    n < 0 && throw(DomainError(n))

    n == 0 && return []

    n == 1 && return [[1]]

    r = triangle(n - 1)
    push!(r, generate_row(r[end]))

    return r

end

function generate_row(previous_row)

    row0 = append!([0], copy(previous_row))

    row1 = append!(copy(previous_row), [0])

    return [ a + b for (a, b) in zip(row0, row1)]
end
